tests/seed_isort_config_test.py

```diff
@@ -12,12 +13,36 @@
 def test_third_party_re():
     assert THIRD_PARTY_RE.search('[isort]\nknown_third_party=cfgv\n')
     assert THIRD_PARTY_RE.search('[isort]\nknown_third_party = cfgv\n')
+    assert THIRD_PARTY_RE.search('[isort]\nknown_third_party\t=\tcfgv\n')
     assert not THIRD_PARTY_RE.search('[isort]\nknown_stdlib = os\n')
     # make sure whitespace isn't greedily matched
     matched = THIRD_PARTY_RE.search('known_third_party=\n').group()
     assert matched == 'known_third_party='
 
 
+def test_known_other_re():
+    isort_config = """
+[settings]
+known_first_party =
+known_third_party =dummy_thirdparty1,dummy_thirdparty2
+"""
+    assert KNOWN_OTHER_RE.search(isort_config)
+    assert KNOWN_OTHER_RE.search(isort_config.replace(" =", "="))
+    assert KNOWN_OTHER_RE.search(isort_config.replace(" =", " = "))
+    assert KNOWN_OTHER_RE.search(isort_config.replace(" =", "\t=\t"))
+    # make sure whitespace isn't greedily matched
+    negative_match = KNOWN_OTHER_RE.search(isort_config)
+    dummy_firstparty_module = "dummy_firstparty1"
+    positive_match = KNOWN_OTHER_RE.search(
+        isort_config.replace(
+            "known_first_party =",
+            "known_first_party=" + dummy_firstparty_module,
+        ),
+    )
+    assert negative_match.group(2) == ""
+    assert positive_match.group(2) == dummy_firstparty_module
```

Comment:
these tests are good, but a little hard to follow -- I tend to try and avoid logic (replace / format / etc.) [in tests](https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html) as it makes it hard to see what's happening here

it might make more sense to have a few targeted tests with clear inputs and outputs

---

though honestly I would have just taken your original reproduction and make sure that works as an integration test :man_shrugging:

Reply:
I started out with one of the integration tests, but quickly realized that the actual problem could be distilled down to "the regex didn't behave properly", so the test needed was for the regex itself.

With a multiline string input, it makes no sense to repeat the multiline string for every permutation, and using a `@pytest.mark.parametrize` annotation seemed to be overkill and separate the permutations even more from the assertions.

The approach I took with `.replace(...)` is one that applies DRY, shows what changes in the permutations, and reduces human error in crafting/modifying tests.

I'm of the opinion that splitting the test up or using `@pytest.mark.parametrize` would only add unnecessary complexity and further obscure the differences between each `assert`.

For example, this seems more complex to me...

```python
@pytest.mark.parametrize(
    ("equal_sign", "firstparty_module"),
    (
        ("=", ""),
        (" =", ""),
        (" = ", ""),
        ("\t=\t", ""),
        ("=", "dummy_firstparty1"),
        (" =", "dummy_firstparty1"),
        (" = ", "dummy_firstparty1"),
        ("\t=\t", "dummy_firstparty1"),
    ),
)
def test_known_other_re(equal_sign, firstparty_module):
    isort_config = """
[settings]
known_first_party{equal_sign}{firstparty_module}
known_third_party{equal_sign}dummy_thirdparty1,dummy_thirdparty2
""".format(equal_sign=equal_sign, firstparty_module=firstparty_module)
    assert KNOWN_OTHER_RE.search(isort_config)
    assert KNOWN_OTHER_RE.search(isort_config).group(2) == firstparty_module
```

... but perhaps it works better for you?

I don't see any other way to split the test without a lot of unnecessary repetition.

Reply:
This has less repetition, but is even more abstracted away from the actual test:

```python
@pytest.mark.parametrize(
    ("equal_sign", "firstparty_module"),
    [
        (equal_sign, firstparty_module)
        for equal_sign in ("=", " =", " = ", "\t=\t")
        for firstparty_module in ("", "dummy_firstparty1")
    ]
)
def test_known_other_re(equal_sign, firstparty_module):
    isort_config = """
[settings]
known_first_party{equal_sign}{firstparty_module}
known_third_party{equal_sign}dummy_thirdparty1,dummy_thirdparty2
""".format(equal_sign=equal_sign, firstparty_module=firstparty_module)
    assert KNOWN_OTHER_RE.search(isort_config)
    assert KNOWN_OTHER_RE.search(isort_config).group(2) == firstparty_module
```

Reply:
DRY in tests is an antipattern as far as I'm concerned -- tests should be easy to read and obvious at a glance what behaviour is being tested (without indirection)

here's how I'd write this test:

```python
@pytest.mark.parametrize(
    ('s', 'expected_groups'),
    (
        ('[isort]\nknown_third_party=\n', ('', '')),
        ('[isort]\nknown_third_party = foo\n', (' ', ' ')),
        ('[isort]\nknown_third_party\t=\tfoo\n', ('\t', '\t')),
        ('[isort]\nknown_third_party =\nknown_wat=wat\n', (' ', '')),
    ),
)
def test_known_third_party_re(s, expected_groups):
    match = THIRD_PARTY_RE.search(s)
    assert match
    assert match.groups() == expected_groups


@pytest.mark.parametrize(
    ('s', 'expected_groups'),
    (
        ('[isort]\nknown_other=\n', ('other', '')),
        ('[isort]\nknown_other = foo\n', ('other', 'foo')),
        ('[isort]\nknown_other\t=\tfoo\n', ('other', 'foo')),
        ('[isort]\nknown_other =\nknown_third_party=wat\n', ('other', '')),
    ),
)
def test_known_other_re(s, expected_groups):
    match = KNOWN_OTHER_RE.search(s)
    assert match
    assert match.groups() == expected_groups
```

clear, concise, easy to see where the inputs and outputs are, the two regexes are tested separately

Reply:
I see what you're saying.  If that's what works for you, we'll do that.  I've filled out the gamut of the parameterization a bit more so that it's clear what the gamut of the variations is.